### PR TITLE
Add GET operations for DeviceGroups and Network Slices

### PR DESCRIPTION
**Pull request by ghislainbourgeois**
This PR goal is to introduce GET operations to the API for DeviceGroups and Network Slices. To do so, the following changes were needed:

1. Move the snapshots from in-memory maps to Mongo in new collections
2. Add GET operations for DeviceGroups and Slices, looking up in those new Mongo collections
3. Add some unit tests for the new code
4. Add test target to the Makefile
5. Add GitHub workflow to run the unit tests

The larger goal is to be able to create a simple web application to configure SD-Core without requiring a full Aether installation for simple use cases.

### CODE REVIEW

**Comment by thakurajayL** [comment]
add to whiltelist

**Comment by thakurajayL** [comment]
ok to test

**Comment by ghislainbourgeois** [comment]
@thakurajayL my org has now signed the CLA, this should be ready for review, sorry it took that long.

**Comment by thakurajayL** [comment]
add to whitelist

**Comment by thakurajayL** [comment]
ok to test

File: Dockerfile [hunk 1]
```diff
@@ -3,7 +3,7 @@
 # SPDX-License-Identifier: Apache-2.0
 #
 
-FROM golang:1.16.0-stretch AS builder
+FROM golang:1.16.0-buster AS builder
```
[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply Dockerfile:6]: Stretch is debian 9, which is an archived release under Extended LTS support. Buster is Debian 10, and it is under regular LTS support. The current stable release would be Bookworm, or Debian 12. I will test for further updates with golang updates that do not require extensive changes.

[Debian releases](https://www.debian.org/releases/)

[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply Dockerfile]: I have updated to latest go version on latest debian version.

File: configmodels/model_device_groups.go [hunk 2]
```diff
@@ -15,9 +15,9 @@
 package configmodels
 
 type DeviceGroups struct {
-	DeviceGroupName string `json:group-name,omitempty"`
+	DeviceGroupName string
```
[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply configmodels/model_device_groups.go]: I don't remember why, but I reverted it and updated the tests.

File: .github/workflows/run-webconsole-unit-tests.yml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2022 Intel Corporation
```
[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply .github/workflows/run-webconsole-unit-tests.yml]: I added the copyright.

File: configapi/api_default.go [hunk 4]
```diff
@@ -15,13 +15,57 @@
...
+	c.JSON(http.StatusOK, deviceGroups)
+}
+
+// GetDeviceGroupsByName -
+func GetDeviceGroupByName(c *gin.Context) {
+	setCorsHeader(c)
+	logger.WebUILog.Infoln("Get Device Group by name")
+
+	var deviceGroup configmodels.DeviceGroups
+	filter := bson.M{"DeviceGroupName": c.Param("group-name")}
```
[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply configapi/api_default.go:58]: `group-name` is a path parameter and is defined in the router. There is no way to get in this function if the param is not there.

File: configapi/api_default.go [hunk 5]
```diff
@@ -58,6 +102,37 @@ func DeviceGroupGroupNamePost(c *gin.Context) {
...
+	c.JSON(http.StatusOK, networkSlices)
+}
+
+// GetNetworkSliceByName -
+func GetNetworkSliceByName(c *gin.Context) {
+	setCorsHeader(c)
+	logger.WebUILog.Infoln("Get Network Slice by name")
+
+	var networkSlice configmodels.Slice
+	filter := bson.M{"SliceName": c.Param("slice-name")}
```
[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply configapi/api_default.go:125]: It is also a path parameter defined in the router.

File: configapi/api_default.go [hunk 6]
```diff
@@ -58,6 +102,37 @@ func DeviceGroupGroupNamePost(c *gin.Context) {
...
+	rawNetworkSlice := RestfulAPIGetOne(sliceDataColl, filter)
+	json.Unmarshal(mapToByte(rawNetworkSlice), &networkSlice)
+
+	if networkSlice.SliceName == "" {
+		c.JSON(http.StatusNotFound, nil)
+	} else {
+		c.JSON(http.StatusOK, networkSlice)
```
[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply configapi/api_default.go:132]: Sure, in the case that the slice does not exist, we get:
```
curl -v ${WEBUI_IP}:5000/config/v1/network-slice/default
*   Trying 10.1.213.205:5000...
* Connected to 10.1.213.205 (10.1.213.205) port 5000 (#0)
> GET /config/v1/network-slice/default HTTP/1.1
> Host: 10.1.213.205:5000
> User-Agent: curl/7.88.1
> Accept: */*
>
< HTTP/1.1 404 Not Found
< Access-Control-Allow-Credentials: true
< Access-Control-Allow-Headers: Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, Authorization, accept, origin, Cache-Control, X-Requested-With
< Access-Control-Allow-Methods: POST, OPTIONS, GET, PUT, PATCH, DELETE
< Access-Control-Allow-Origin: *
< Content-Type: application/json; charset=utf-8
< Date: Mon, 11 Sep 2023 20:57:59 GMT
< Content-Length: 4
<
* Connection #0 to host 10.1.213.205 left intact
null
```

In the case where it exists, we get this:
```
curl -v ${WEBUI_IP}:5000/config/v1/network-slice/default
*   Trying 10.1.213.205:5000...
* Connected to 10.1.213.205 (10.1.213.205) port 5000 (#0)
> GET /config/v1/network-slice/default HTTP/1.1
> Host: 10.1.213.205:5000
> User-Agent: curl/7.88.1
> Accept: */*
>
< HTTP/1.1 200 OK
< Access-Control-Allow-Credentials: true
< Access-Control-Allow-Headers: Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, Authorization, accept, origin, Cache-Control, X-Requested-With
< Access-Control-Allow-Methods: POST, OPTIONS, GET, PUT, PATCH, DELETE
< Access-Control-Allow-Origin: *
< Content-Type: application/json; charset=utf-8
< Date: Mon, 11 Sep 2023 20:59:34 GMT
< Content-Length: 238
<
* Connection #0 to host 10.1.213.205 left intact
{"SliceName":"default","slice-id":{"sst":"1","sd":"010203"},"site-device-group":["cows"],"site-info":{"site-name":"demo","plmn":{"mcc":"208","mnc":"93"},"gNodeBs":[{"name":"demo-gnb1","tac":1}],"upf":{"upf-name":"upf","upf-port":"8805"}}}
```

File: configmodels/model_device_groups.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply configmodels/model_device_groups.go:18]: I reverted that commit again, it was causing some issues. I am investigating to have a proper fix. I will leave this thread open until then.

[the remark this answers is not in the document]
Comment (ghislainbourgeois) [reply configmodels/model_device_groups.go]: I found the issue and added more changes to keep `group-name` as the field name in JSON.

**Comment by thakurajayL** [comment]
add to whitelist

**Comment by thakurajayL** [comment]
ok to test

**Comment by thakurajayL** [comment]
retest this please

**Review by thakurajayL — APPROVED** [review]

**Pull request merged**